### PR TITLE
Initial version for HtmlComment plugin

### LINKED ISSUE #9662: HTML comments: Conversion upcast
## Provide a description of the task

Handle HTML comments upcast (`<$comment>` with a custom property `$rawContent` to a marker).

### LINKED ISSUE #9663: HTML comments: Conversion downcast
## Provide a description of the task

Handle HTML comments downcast (marker to `<$comment>` with a custom property `$rawContent`).

### LINKED ISSUE #10110: HTML comments: Add unit tests for `HtmlComment` plugin
## Provide a description of the task

Add unit tests for `HtmlComment` plugin.

### LINKED ISSUE #10136: Implement an initial version of HTML comments support
Scope: Cleaning up the work that has already been done in #8822 and polishing #10109.

### PR DESCRIPTION
**Pull request by psmyrek**
### Suggested merge commit message ([convention](https://ckeditor.com/docs/ckeditor5/latest/framework/guides/contributing/git-commit-message-convention.html))

Internal (html-support): Initial version of the `HtmlComment` plugin. Closes #10136. Closes #10110. Closes #9663. Closes #9662.

---

This is the initial version of the `HtmlComment` just to have anything useful in the feature branch. There are still some problems to be solved and missing things to be added. See epic link: https://app.zenhub.com/workspaces/ckeditor-5-5d9c8a4eac250d0001664feb/issues/ckeditor/ckeditor5/8822.

### CODE REVIEW

File: packages/ckeditor5-engine/src/view/domconverter.js [hunk 1]
```diff
@@ -237,8 +238,9 @@ export default class DomConverter {
 					this.bindDocumentFragments( domElement, viewNode );
 				}
 			} else if ( viewNode.is( 'uiElement' ) ) {
-				// UIElement has its own render() method (see #799).
```
Comment (ma2ciek) [inline packages/ckeditor5-engine/src/view/domconverter.js:240]: I guess it's still true for elements other than `$comment`

File: packages/ckeditor5-engine/src/view/domconverter.js [hunk 2]
```diff
@@ -1298,6 +1302,35 @@ export default class DomConverter {
...
+	}
+
+	/**
+	 * Checks if view element's content should be treated as a raw data.
+	 *
+	 * @param {Element} viewElement View element to check.
+	 * @param {Object} options Conversion options. See {@link module:engine/view/domconverter~DomConverter#domToView} options parameter.
+	 * @returns {Boolean}
+	 */
+	_isViewElementWithRawContent( viewElement, options ) {
```
Comment (ma2ciek) [inline packages/ckeditor5-engine/src/view/domconverter.js:1331]: Missing `@private` tag

File: packages/ckeditor5-engine/tests/view/domconverter/dom-to-view.js [hunk 3]
```diff
@@ -171,10 +172,30 @@ describe( 'DomConverter', () => {
...
+
+			expect( viewComment ).to.be.an.instanceof( ViewUIElement );
+			expect( viewComment.name ).to.equal( '$comment' );
+
+			expect( viewComment.getCustomProperty( '$rawContent' ) ).to.equal( 'abc' );
+
+			expect( converter.mapViewToDom( viewComment ) ).to.not.equal( domComment );
+		} );
+
+		it( 'should create UIElement for comment and bind element', () => {
```
Comment (ma2ciek) [inline packages/ckeditor5-engine/tests/view/domconverter/dom-to-view.js:188]:
```suggestion
		it( 'should create UIElement for comment and bind elements', () => {
```

File: packages/ckeditor5-engine/tests/view/domconverter/view-to-dom.js [hunk 4]
```diff
@@ -186,6 +187,34 @@ describe( 'DomConverter', () => {
 			expect( domSvg.createSVGRect ).to.be.a( 'function' );
 		} );
 
+		it( 'should create DOM comment node from view element', () => {
```
Comment (ma2ciek) [inline packages/ckeditor5-engine/tests/view/domconverter/view-to-dom.js:190]:
```suggestion
		it( 'should create a DOM comment node from a view `$comment` UIElement', () => {
```

File: packages/ckeditor5-engine/tests/view/domconverter/view-to-dom.js [hunk 5]
```diff
@@ -186,6 +187,34 @@ describe( 'DomConverter', () => {
...
+			const domComment = converter.viewToDom( viewComment, document );
+
+			expect( domComment ).to.be.an.instanceof( Comment );
+			expect( domComment.nodeName ).to.equal( '#comment' );
+			expect( domComment.data ).to.equal( 'foo' );
+
+			expect( converter.mapDomToView( domComment ) ).to.not.equal( viewComment );
+		} );
+
+		it( 'should create DOM comment node from view element and bind element', () => {
```
Comment (ma2ciek) [inline packages/ckeditor5-engine/tests/view/domconverter/view-to-dom.js:204]:
```suggestion
		it( 'should create a DOM comment node from a view `$comment` UIElement and bind them', () => {
```

File: packages/ckeditor5-engine/src/view/domconverter.js (hunk 1 above)
Comment (psmyrek) [reply packages/ckeditor5-engine/src/view/domconverter.js:240]: Rewritten a bit to restore this deleted comment.

**Comment by psmyrek** [comment]
Important question: may I remove the deprecation warning from `UpcastHelpers#elementToMarker()`, as the current solution for handling HTML comments is based on this helper?

File: packages/ckeditor5-html-support/tests/htmlcomment.js [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+		return editor.destroy();
+	} );
+
+	it( 'should be named', () => {
+		expect( HtmlComment.pluginName ).to.equal( 'HtmlComment' );
+	} );
+
+	it( 'should be loaded', () => {
+		expect( editor.plugins.get( HtmlComment ) ).to.be.instanceOf( HtmlComment );
+	} );
```
Comment (ma2ciek) [inline packages/ckeditor5-html-support/tests/htmlcomment.js:44]: I know that these tests are present in all our plugins but they are quite dumb TBH :smile: 

Perhaps we could replace these two tests with the following: (?)

```
it( 'should be loadable using its plugin name', () => {
	expect( editor.plugins.get( 'HtmlComment' ) ).to.be.instanceOf( HtmlComment );
} );
```

File: packages/ckeditor5-html-support/tests/htmlcomment.js [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+		model.change( writer => {
+			writer.addMarker( name, {
+				usingOperation: true,
+				affectsData: true,
+				range: writer.createRange(
+					writer.createPositionAt( element, offset )
+				)
+			} );
+
+			writer.setAttribute( name, content, root );
```
Comment (ma2ciek) [inline packages/ckeditor5-html-support/tests/htmlcomment.js:181]: Could we wrap this in a separate method? It's not a part of adding markers.

File: packages/ckeditor5-html-support/tests/htmlcomment.js [hunk 8]
```diff
@@ -0,0 +1,184 @@
+/**
```
Comment (ma2ciek) [inline packages/ckeditor5-html-support/tests/htmlcomment.js:1]: I'd add some tests with comments at the edge of the content (one for upcast, downcast, and post-fixer - even if this does not work currently as expected).

File: packages/ckeditor5-html-support/tests/htmlcomment.js (hunk 6 above)
Comment (psmyrek) [reply packages/ckeditor5-html-support/tests/htmlcomment.js]: Yeah, good idea.
Changed as proposed.

File: packages/ckeditor5-html-support/tests/htmlcomment.js (hunk 7 above)
Comment (psmyrek) [reply packages/ckeditor5-html-support/tests/htmlcomment.js:257]: Done.
I added symmetrical methods: `_getCommentContent( markerName )` and `_setCommentContent( markerName, commentContent )`.

File: packages/ckeditor5-html-support/tests/htmlcomment.js (hunk 8 above)
Comment (psmyrek) [reply packages/ckeditor5-html-support/tests/htmlcomment.js:1]: Done.
After adding new tests, one is failing currently.

File: packages/ckeditor5-html-support/tests/htmlcomment.js (hunk 7 above)
Comment (psmyrek) [reply packages/ckeditor5-html-support/tests/htmlcomment.js]: I got it wrong. The private methods mentioned in my comment above are not needed (and now are removed). Changes were needed in test file, where the `addMarker()` function actually was doing 2 things: adding a marker and setting a root attribute.

Now `addMarker()` adds a marker only, and root attribute is added separately.

File: packages/ckeditor5-html-support/tests/htmlcomment.js [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+			// Currently, there are 2 $root attributes associated with markers.
+			expect( rootAttributes ).to.have.length( 0 );
+
+			const markers = [ ...editor.model.markers ].filter( marker => marker.name.startsWith( '$comment' ) );
+
+			// Currently, there are 2 markers associated with comment nodes.
+			expect( markers ).to.have.length( 0 );
+		} );
+
+		it( 'should remove markers and their corresponding $root attributes representing removed comments only', () => {
```
Comment (ma2ciek) [inline packages/ckeditor5-html-support/tests/htmlcomment.js:203]:
```suggestion
		it( 'should remove comment markers and their corresponding root attributes when the content including comments is removed', () => {
```

File: packages/ckeditor5-html-support/tests/htmlcomment.js [hunk 10]
```diff
@@ -0,0 +1,236 @@
...
+
+			addMarker( '$comment:2', paragraph, 2 );
+			root._setAttribute( '$comment:2', ' comment 2 ' );
+
+			expect( editor.getData() ).to.equal( '<div><div><div><p>F<!-- comment 1 -->o<!-- comment 2 -->o</p></div></div></div>' );
+		} );
+	} );
+
+	describe( 'marker removal post fixer', () => {
+		it( 'should remove all markers and their corresponding $root attributes', () => {
```
Comment (ma2ciek) [inline packages/ckeditor5-html-support/tests/htmlcomment.js:161]:
```suggestion
				it( 'should remove comment markers and their corresponding root attributes when the whole content is removed', () => {
```

File: packages/ckeditor5-html-support/tests/htmlcomment.js [hunk 11]
```diff
@@ -0,0 +1,236 @@
...
+
+			const markers = [ ...editor.model.markers ].filter( marker => marker.name.startsWith( '$comment' ) );
+
+			expect( markers ).to.have.length( 0 );
+		} );
+
+		// Currently, this test fails. Removing all content from the editor does not remove markers located at $root's boundary.
+		// Since the markers survive, so do comments and their $root attributes.
+		// See https://github.com/ckeditor/ckeditor5/issues/10117.
+		it.skip( 'should remove all markers located at $root\'s boundary and their corresponding $root attributes', () => {
```
Comment (ma2ciek) [inline packages/ckeditor5-html-support/tests/htmlcomment.js:182]:
```suggestion
		it( 'should remove all boundary comment markers and their corresponding root attributes when the whole content is removed', () => {
```

File: packages/ckeditor5-html-support/tests/htmlcomment.js [hunk 12]
```diff
@@ -0,0 +1,236 @@
...
+				const firstParagraph = root.getChild( 0 );
+
+				writer.remove( writer.createRangeOn( firstParagraph ) );
+			} );
+
+			const rootAttributes = [ ...root.getAttributeKeys() ].filter( attr => attr.startsWith( '$comment' ) );
+
+			expect( rootAttributes ).to.have.length( 2 );
+
+			const markers = [ ...editor.model.markers ].filter( marker => marker.name.startsWith( '$comment' ) );
```
Comment (ma2ciek) [inline packages/ckeditor5-html-support/tests/htmlcomment.js:216]: I'd change all `markers` to `commentMarkers` in these tests.

Comment (psmyrek) [reply packages/ckeditor5-html-support/tests/htmlcomment.js]: Done.

File: packages/ckeditor5-html-support/tests/htmlcomment.js (hunk 11 above)
Comment (psmyrek) [reply packages/ckeditor5-html-support/tests/htmlcomment.js]: Slightly changed to fit in one line.

File: packages/ckeditor5-html-support/tests/htmlcomment.js (hunk 10 above)
Comment (psmyrek) [reply packages/ckeditor5-html-support/tests/htmlcomment.js]: Slightly changed to fit in one line.

File: packages/ckeditor5-html-support/tests/htmlcomment.js (hunk 9 above)
Comment (psmyrek) [reply packages/ckeditor5-html-support/tests/htmlcomment.js]: Done.

**Comment by ma2ciek** [comment]
All looks good to me now.

We need to make a decision regarding the deprecated upcast method -  #10128 as it blocks this PR now.
